From 2d7da45180e602ab920578714546fd78b5093bb6 Mon Sep 17 00:00:00 2001
From: Jon Nettleton <jon@solid-run.com>
Date: Tue, 8 Feb 2022 15:22:12 +0100
Subject: [PATCH 07/11] regulator: pca9450: Add dt property pca,i2c-lt-en

The PCA9450 (at least the DS version) has a built in level
translator that allows I2C lines to be shifted from 3.3V to 1.8V.
This is not a regulator, or a gate just a level shifter that can
be configured to be enabled in various run modes.  It is more of
a power management feature for the PMIC.  By default this is
disabled which means unless the firmware enabled the level
translator there is no communication possible with the devices
on the I2C bus.

Rather than doing a bunch of fancy work to make this work as a
regulator...because it isn't...we just add a property so this
function of the PMIC can be configured properly at probe time.

Signed-off-by: Jon Nettleton <jon@solid-run.com>
---
 drivers/regulator/pca9450-regulator.c | 7 +++++++
 include/linux/regulator/pca9450.h     | 2 ++
 2 files changed, 9 insertions(+)

diff --git a/drivers/regulator/pca9450-regulator.c b/drivers/regulator/pca9450-regulator.c
index d59305a6dd41..a6a594c27643 100644
--- a/drivers/regulator/pca9450-regulator.c
+++ b/drivers/regulator/pca9450-regulator.c
@@ -716,6 +716,7 @@ static int pca9450_i2c_probe(struct i2c_client *i2c,
 	struct regulator_config config = { };
 	struct pca9450 *pca9450;
 	unsigned int device_id, i;
+	uint8_t i2c_lt_en;
 	int ret;
 
 	if (!i2c->irq) {
@@ -836,6 +837,12 @@ static int pca9450_i2c_probe(struct i2c_client *i2c,
 		return ret;
 	}
 
+	ret = of_property_read_u8(i2c->dev.of_node, "pca,i2c-lt-en", &i2c_lt_en);
+	if (!ret) {
+		regmap_update_bits(pca9450->regmap, PCA9450_REG_CONFIG2,
+					CFG2_I2C_LT_MASK, i2c_lt_en);
+        }
+
 	dev_info(&i2c->dev, "%s probed.\n",
 		type == PCA9450_TYPE_PCA9450A ? "pca9450a" : "pca9450bc");
 
diff --git a/include/linux/regulator/pca9450.h b/include/linux/regulator/pca9450.h
index 71902f41c919..1ddc303dfb5b 100644
--- a/include/linux/regulator/pca9450.h
+++ b/include/linux/regulator/pca9450.h
@@ -226,4 +226,6 @@ enum {
 #define WDOG_B_CFG_COLD_LDO12		0x80
 #define WDOG_B_CFG_COLD			0xC0
 
+#define CFG2_I2C_LT_MASK		0x03
+
 #endif /* __LINUX_REG_PCA9450_H__ */
-- 
2.35.1

